File: desktop/desktop_app/desktop_app/Models/IncidentModel.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace desktop_app.Models
{
    public class IncidentModel
    {
        [JsonPropertyName("_id")]
        public string Id { get; set; } = "";

        [JsonPropertyName("room_id")]
        public string RoomId { get; set; } = "";

        /// <summary>Algunas respuestas JSON usan camelCase en lugar de snake_case.</summary>
        [JsonPropertyName("roomId")]
        public string? RoomIdCamel { get; set; }

        [JsonIgnore]
        public string RoomNumber { get; set; } = "";

        [JsonPropertyName("type")]
        public string Type { get; set; } = "";

        [JsonPropertyName("severity")]
        public string Severity { get; set; } = "";

        [JsonPropertyName("description")]
        public string Description { get; set; } = "";

        [JsonPropertyName("status")]
        public string Status { get; set; } = "";

        [JsonPropertyName("reported_by")]
        public string ReportedBy { get; set; } = "";

        [JsonPropertyName("assigned_to")]
        public string? AssignedTo { get; set; }

        [JsonPropertyName("assignedTo")]
        public string? AssignedToCamel { get; set; }

        /// <summary>Texto para grid (nombre o «Sin asignar»); se rellena al listar.</summary>
        [JsonIgnore]
        public string AssignedToDisplay { get; set; } = "Sin asignar";

        [JsonPropertyName("reported_at")]
        public DateTime ReportedAt { get; set; }

        [JsonPropertyName("resolved_at")]
        public DateTime? ResolvedAt { get; set; }

        [JsonPropertyName("internal_notes")]
        public List<IncidentNote> InternalNotes { get; set; } = new();

        [JsonPropertyName("history")]
        public List<IncidentHistoryEntry> History { get; set; } = new();

        [JsonIgnore]
        public string SeverityLabel => Severity switch
        {
            "low" => "Baja",
            "medium" => "Media",
            "high" => "Alta",
            _ => Severity
        };

        /// <summary>Hay empleado asignado (JSON puede venir en snake_case o camelCase).</summary>
        [JsonIgnore]
        public bool HasAssignee =>
            !string.IsNullOrWhiteSpace(AssignedTo) || !string.IsNullOrWhiteSpace(AssignedToCamel);

        [JsonIgnore]
        public string StatusLabel
        {
            get
            {
                if (string.Equals(Status, "resolved", StringComparison.OrdinalIgnoreCase))
                    return "Resuelta";
                if (HasAssignee)
                    return "En proceso";
                if (string.Equals(Status, "open", StringComparison.OrdinalIgnoreCase))
                    return "Comunicado";
                return Status;
            }
        }
    }

    public class IncidentNote
    {
        [JsonPropertyName("note")]
        public string Note { get; set; } = "";

        [JsonPropertyName("author_id")]
        public string AuthorId { get; set; } = "";

        [JsonPropertyName("created_at")]
        public DateTime CreatedAt { get; set; }

        /// <summary>Rellenado en cliente a partir de <see cref="AuthorId"/> (no viene del JSON).</summary>
        [JsonIgnore]
        public string AuthorLabel { get; set; } = "";
    }

    public class IncidentHistoryEntry
    {
        [JsonPropertyName("action")]
        public string Action { get; set; } = "";

        [JsonPropertyName("by")]
        public string? By { get; set; }

        [JsonPropertyName("at")]
        public DateTime At { get; set; }

        /// <summary>Nombre mostrable; se rellena en cliente desde <see cref="By"/>.</summary>
        [JsonIgnore]
        public string ByLabel { get; set; } = "";
    }

    public class IncidentsResponse
    {
        [JsonPropertyName("items")]
        public List<IncidentModel> Items { get; set; } = new();
    }
}

File: api/src/models/incidentModel.js
import { Schema, Types, model } from "mongoose";

/**
 * @typedef {Object} Incident
 * @property {import("mongoose").Types.ObjectId} _id
 * @property {import("mongoose").Types.ObjectId} room_id
 * @property {"ruido"|"limpieza"|"averia"|"otro"} type
 * @property {"low"|"medium"|"high"} severity
 * @property {string} description
 * @property {"open"|"resolved"} status
 * @property {import("mongoose").Types.ObjectId} reported_by
 * @property {import("mongoose").Types.ObjectId|null} assigned_to
 * @property {{ note: string, author_id: import("mongoose").Types.ObjectId, created_at: Date }[]} internal_notes
 * @property {{ action: string, by: import("mongoose").Types.ObjectId|null, at: Date, details?: any }[]} history
 * @property {Date} reported_at
 * @property {Date|null} resolved_at
 */
const incidentSchema = new Schema({
  room_id: { type: Types.ObjectId, ref: "room", required: true, index: true },
  type: {
    type: String,
    enum: ["ruido", "limpieza", "averia", "otro"],
    required: true
  },
  severity: {
    type: String,
    enum: ["low", "medium", "high"],
    required: true
  },
  description: { type: String, required: true, trim: true, maxlength: 2000 },
  status: { type: String, enum: ["open", "resolved"], default: "open", index: true },
  reported_by: { type: Types.ObjectId, ref: "user", required: true },
  assigned_to: { type: Types.ObjectId, ref: "user", default: null, index: true },
  internal_notes: {
    type: [
      {
        note: { type: String, required: true, trim: true, maxlength: 2000 },
        author_id: { type: Types.ObjectId, ref: "user", required: true },
        created_at: { type: Date, default: Date.now }
      }
    ],
    default: []
  },
  history: {
    type: [
      {
        action: { type: String, required: true },
        by: { type: Types.ObjectId, ref: "user", default: null },
        at: { type: Date, default: Date.now },
        details: { type: Schema.Types.Mixed, default: null }
      }
    ],
    default: []
  },
  reported_at: { type: Date, default: Date.now, index: true },
  resolved_at: { type: Date, default: null }
});

incidentSchema.index({ room_id: 1, reported_at: -1 });
incidentSchema.index({ room_id: 1, status: 1, severity: 1, reported_at: -1 });
incidentSchema.index({ status: 1, severity: 1, reported_at: -1 });

export const incidentModel = model("incident", incidentSchema);

